src/shadow-embed.js

```diff
@@ -314,6 +316,40 @@ function getStylesheetRules(doc, css) {
   }
 }
 
+/**
+ * @param {!ShadowRoot} shadowRoot
+ * @param {string} name
+ * @param {string} cssText
+ */
+export function installShadowStyle(shadowRoot, name, cssText) {
+  const doc = shadowRoot.ownerDocument;
+  const win = toWin(doc.defaultView);
+  if (shadowRoot.adoptedStyleSheets !== undefined &&
```

Comment:
Do we have any sense for the performance difference? I'm guessing it becomes more significant when the cache starts being used (2+ of the same component type)

Reply:
Due to some nuances in the spec, this should be faster even for one component. Depending on the size of the stylesheet, it's "significantly faster" for 2+ components. The optimization goes beyond reusing the same memory structures. The adopted stylesheets are applied synchronously and the correct style tree can be created at the same time as the DOM is constructed. I.e. no re-layout after stylesheet has been fully parsed. But more importantly: this is a way for us to have a centralized way to push stylesheets for components. We have something similar for AMP elements, and this gives us the path for the shadow components as well.

Reply:
For posterity: one-to-one comparison of adopted stylesheets vs `<style>` are fairly complicated because the adopted stylehseet API captures the construction latency much more accurately than a single `<style>` tag due to parse/style-tree construction. I'm talking to benchmark folks to get a better picture.